data-structures/src/data/structure/linkedlist/LinkedList.java

```diff
@@ -0,0 +1,239 @@
+package data.structure.linkedlist;
+import static  java.util.Objects.*;
+
+public class LinkedList<E> {
+    private class Node {
+        E data;
+        Node next;
+        Node previous;
+
+        Node(final E data) {
+            this.data = data;
+            this.next = null;
+            this.previous = null;
+        }
+    }
+
+    private Node head;
+    private Node tail;
+    private int size;
+
+    public LinkedList() {
+        this.head = null;
+        this.tail = null;
+        size = 0;
+    }
+
+    public void add(E element) {
+        requireNonNull(element, "The key must not be null!");
+
+        if (isNull(head)) {
+            head = new Node(element);
+            tail = head;
+            size++;
+
+        } else {
+            tail.next = new Node(element);
+            final Node temp = tail;
+            tail = tail.next;
+            tail.previous = temp;
+            size++;
+        }
+    }
+
+    public void add(int index, E element) {
+        requireNonNull(element, "The element must not be null!");
+
+        if (index > size || index < 0) {
+            throw new IllegalArgumentException("Enter a proper index value!");
+        }
+
+        if (index == size) {
+            add(element);
+            return;
+        }
+
+        Node requiredNode = new Node(element);
+
+        if (index == 0) {
+            head.previous = requiredNode;
+            requiredNode.next = head;
+            head = requiredNode;
+            size++;
+            return;
+        }
+
+        Node requestedPlace = findNode(index);
+
+        requestedPlace.previous.next = requiredNode;
+        requiredNode.previous = requestedPlace.previous;
+        requiredNode.next = requestedPlace;
+        requestedPlace.previous = requiredNode;
+    }
+
+    public void addAll(LinkedList<E> other) {
+        requireNonNull(other, "The list must not be null!");
+
+        LinkedList<E> copy = makeCopyList(other);
+
+        tail.next = copy.head;
+        copy.head.previous = tail;
+        tail = copy.tail;
+        size += copy.size;

```

Comment:
Awesome job, very well done! 💯 Copying the other list was the key move here. The chaining at the end is just the cherry on top.